src/adapters/adbutler.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * @overview AdButler Prebid.js adapter.
+ * @author dkharton
+ */
+
+'use strict';
+
+var utils = require('../utils.js');
+var adloader = require('../adloader.js');
+var bidmanager = require('../bidmanager.js');
+var bidfactory = require('../bidfactory.js');
+
+var AdButlerAdapter = function AdButlerAdapter() {
+  
+  function _callBids(params) {
+    
+    var bids = params.bids || [],
+        callbackData = {},
+        zoneCount = {},
+        pageID = Math.floor(Math.random() * 10e6);
+
+    //Build and send bid requests
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i],
+        zoneID = utils.getBidIdParamater('zoneID', bid.params),
+        callbackID;  
+        
+      if(!(zoneID in zoneCount)){
+        zoneCount[zoneID] = 0;
+      }
+      
+      //build callbackID to get placementCode later  
+      callbackID = zoneID + '_' +zoneCount[zoneID];
+
+      callbackData[callbackID] = {};
+      callbackData[callbackID].bidId = bid.bidId;
+      
+      var adRequest = buildRequest(bid,zoneCount[zoneID],pageID);
+      zoneCount[zoneID]++;
+
+      adloader.loadScript(adRequest);
+    }
+      
+    //Define callback function for bid responses
+    $$PREBID_GLOBAL$$.adbutlerCB = function(aBResponseObject){
+        
+      var bidResponse = {},
+        callbackID = aBResponseObject.zone_id+'_'+aBResponseObject.place,
+        width = parseInt(aBResponseObject.width),
+        height = parseInt(aBResponseObject.height),
+        isCorrectSize = false,
+        isCorrectCPM = true,
+        CPM,minCPM,maxCPM,
+        bidObj = utils.getBidRequest(callbackData[callbackID].bidId);
```

Comment:
This can cause another `TypeError` which occasionally fails unit tests. Checking `callbackData[callbackID]` before accessing `bidId` will fix this. Something like `bidObj = callbackData[callbackID] ? utils.getBidRequest(callbackData[callbackID].bidId) : null;` will work as a one-liner but feel free to do the check however you prefer